redash/query_runner/google_analytics4.py

```diff
@@ -0,0 +1,183 @@
+import datetime
+import logging
+from base64 import b64decode
+
+import requests
+
+from redash.query_runner import (
+    TYPE_DATE,
+    TYPE_DATETIME,
+    TYPE_FLOAT,
+    TYPE_INTEGER,
+    TYPE_STRING,
+    BaseQueryRunner,
+    register,
+)
+from redash.utils import json_dumps, json_loads
+
+logger = logging.getLogger(__name__)
+
+try:
+    import httplib2
+    from apiclient.discovery import build
+    from oauth2client.service_account import ServiceAccountCredentials
+
+    enabled = True
+except ImportError:
+    enabled = False
+
+types_conv = dict(
+    STRING=TYPE_STRING,
+    INTEGER=TYPE_INTEGER,
+    FLOAT=TYPE_FLOAT,
+    DATE=TYPE_DATE,
+    DATETIME=TYPE_DATETIME,
+)
+
+ga_report_endpoint = "https://analyticsdata.googleapis.com/v1beta/properties/{propertyId}:runReport"
+ga_metadata_endpoint = "https://analyticsdata.googleapis.com/v1beta/properties/{propertyId}/metadata"
+
+
+def format_column_value(column_name, value, columns):
+    column_type = [col for col in columns if col["name"] == column_name][0]["type"]
+
+    if column_type == TYPE_DATE:
+        value = datetime.datetime.strptime(value, "%Y%m%d")
+    elif column_type == TYPE_DATETIME:
+        if len(value) == 10:
+            value = datetime.datetime.strptime(value, "%Y%m%d%H")
+        elif len(value) == 12:
+            value = datetime.datetime.strptime(value, "%Y%m%d%H%M")
+        else:
+            raise Exception("Unknown date/time format in results: '{}'".format(value))
+
+    return value
+
+
+def parse_ga_response(response):
+    columns = []
+    for h in response["dimensionHeaders"]:
+        data_type = None
+        if h["name"] == "date":
+            data_type = "DATE"
+        elif h["name"] == "dateHour":
+            data_type = "DATETIME"
+        columns.append(
+            {
+                "name": h["name"],
+                "friendly_name": h["name"],
+                "type": types_conv.get(data_type, "string"),
+            }
+        )
+
+    for h in response["metricHeaders"]:
+        data_type = None
+        if h["name"] == "date":
+            data_type = "DATE"
+        elif h["name"] == "dateHour":
+            data_type = "DATETIME"
+        columns.append(
+            {
+                "name": h["name"],
+                "friendly_name": h["name"],
+                "type": types_conv.get(data_type, "string"),
+            }
+        )
```

Comment:
Since this is ~identical to L59-71, can we have a nested for loop instead to keep the code more DRY?

Reply:
Added a common method to fix the duplicate code.